index/segment/mem/trigram_terms_dict.go

```diff
@@ -0,0 +1,216 @@
+		if err := t.backingDict.Insert(
+			doc.Field{
+				Name:      field.Name,
+				Value:     tri,
+				ValueType: field.ValueType,
+			},
+			i,
+		); err != nil {
+			return fmt.Errorf("unable to insert trigram %s into backing dictionary: %v", tri, err)
+		}
+	}
+	return nil
+}
+
+func (t *trigramTermsDictionary) Fetch(
+	fieldName []byte,
+	fieldValueFilter []byte,
+	opts termFetchOptions,
+) (segment.PostingsList, error) {
+	filter := string(fieldValueFilter)
+	re, err := syntax.Parse(filter, syntax.Perl)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse query filter %s: %v", filter, err)
+	}
+	q := index.RegexpQuery(re)
+	ids, err := t.postingQuery(fieldName, q, nil, false)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get postings list matching query: %v", err)
+	}
+	return ids, nil
+}
+
+func (t *trigramTermsDictionary) postingQuery(
+	fieldName []byte,
+	q *index.Query,
+	candidateIDs segment.PostingsList,
+	created bool,
+) (segment.PostingsList, error) {
+	switch q.Op {
+	case index.QNone:
+		// Do nothing.
+
+	case index.QAll:
+		if candidateIDs != nil {
+			return candidateIDs, nil
+		}
+		// Match all queries are not supported by the trigram terms dictionary.
+		return nil, errUnsupportedQuery
+
+	case index.QAnd:
+		for _, tri := range q.Trigram {
+			ids, err := t.docIDsForTrigram(fieldName, tri)
+			if err != nil {
+				return nil, err
+			}
+			if ids == nil {
+				return t.opts.PostingsListPool().Get(), nil
+			}
+			if !created {
+				candidateIDs = ids.Clone()
+				created = true
+			} else {
+				candidateIDs.Intersect(ids)
+			}
+			if candidateIDs.IsEmpty() {
+				return candidateIDs, nil
+			}
+		}
+
+		for _, sub := range q.Sub {
+			ids, err := t.postingQuery(fieldName, sub, candidateIDs, created)
+			if err != nil {
+				return nil, err
+			}
+			if ids == nil {
+				return t.opts.PostingsListPool().Get(), nil
+			}
+			if !created {
+				candidateIDs = ids
+				created = true
+			} else {
+				candidateIDs.Intersect(ids)
+			}
+			if candidateIDs.IsEmpty() {
+				return candidateIDs, nil
+			}
+		}
+
+	case index.QOr:
+		for _, tri := range q.Trigram {
+			ids, err := t.docIDsForTrigram(fieldName, tri)
+			if err != nil {
+				return nil, err
+			}
+			if ids == nil {
+				continue
+			}
+			if !created {
+				candidateIDs = ids.Clone()
+				created = true
+			} else {
+				candidateIDs.Union(ids)
+			}
+		}
+
+		for _, sub := range q.Sub {
+			ids, err := t.postingQuery(fieldName, sub, candidateIDs, created)
+			if err != nil {
+				return nil, err
+			}
+			if ids == nil {
+				return t.opts.PostingsListPool().Get(), nil
+			}
+			if !created {
+				candidateIDs = ids
+				created = true
+			} else {
+				candidateIDs.Union(ids)
+			}
+		}
+	}
+
+	return candidateIDs, nil
+}
+
+func (t *trigramTermsDictionary) docIDsForTrigram(
+	fieldName []byte,
+	tri string,
+) (segment.ImmutablePostingsList, error) {
+	return t.backingDict.Fetch(fieldName, []byte(tri), termFetchOptions{isRegexp: false})
```

Comment:
Would it be better to expose a `FetchString()` on the simpleTerms dictionary to avoid the []byte alloc here?

Reply:
That's a fair point. I'll take this as a TODO since I think we may want to evaluate our use of strings vs. slices throughout the package because we do something unnecessary conversions in other places as well.